cli/types/index.d.ts

```diff
@@ -4,7 +4,7 @@
 //                 Mike Woudenberg <https://github.com/mikewoudenberg>
 //                 Robbert van Markus <https://github.com/rvanmarkus>
 //                 Nicholas Boll <https://github.com/nicholasboll>
-// TypeScript Version: 3.4
+// TypeScript Version: 3.7
```

Comment:
This is updated to 3.7. Because `@types/node` uses [assertion functions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions) added at 3.7.

Reply:
Will this change break consumers of the cypress types who are on TS 3.4? If so, we may have to keep `@types/node` at `14` in the `cli`.

Reply:
The only breaking case might be:

* no `@types/node`
* TS <= 3.6

Because in this case, `@types/node` is installed at the root of `node_modules` and it might cause some problems. 

But this is unlikely because:

* TS 3.7 was [released at November 5th, 2019](https://devblogs.microsoft.com/typescript/announcing-typescript-3-7/), about 2.5 years ago. 
* Cypress 9.x requires Node 16.x. It means most users might have installed `@types/node` 16.x on their project. `@types/node` of `cli` is outdated.

If breaking change is the problem, I think it would be safe to include this PR to 10.0 branch if the release isn't delayed.

Reply:
Looks like older TS versions (3.4~3.7) have around 150k-200k downloads a week: https://www.npmjs.com/package/typescript. 

Not entirely sure of our minimum requirements, would definitely be good to have a minimum version we could tell people they need.

Reply:
Cypress actually supports Node 12 and up, it's just that our development is done in 16: https://github.com/cypress-io/cypress/blob/7aa9b20592c03049baf42ab260c666cb5c3f4cc6/cli/package.json#L110-L112

So we're already kind not doing what I expected here, because we're depending on `@types/node@14.14.31` in the `cli`...

> Because in this case, @types/node is installed at the root of node_modules and it might cause some problems.

I see. So this would break a project's types if it relied on `"types": "["node"]"` in `tsconfig.json`, but did not explicitly include `@types/node` in `package.json`, and was using the `@types/node` transitively depended upon by `cypress`. It would bump from 14.14.31 to 16 for their entire project. Is that correct?

Otherwise, things would be fine, their project's types would use their version of `@types/node` and `cypress` would use the version shipped with `cypress`. Are namespace collisions possible?

I wonder if we should have some kind of rules about what version of `@types/node` we depend on in the `cli`, in light of all of this. I'm tempted to say it doesn't matter, since people shouldn't be using our `@types/node` anyways. But at the same time... changing it outside of a breaking release would definitely break whatever % of users are implicitly using our `@types/node`.

Reply:
I made repos to help you understand. 

* https://github.com/sainthkh/ts34 -> It depends on `@types/node@16.x`. 
* https://github.com/sainthkh/tst -> It depends on ts34.

`ts34` is a simplified `cypress` package and `tst` is a user-project. 

----

If there is only one package in the dependency tree, it's installed at the root of `node_modules`. 

In `tst` example, there is only one `@types/node` in the tree. So, it's installed at the root of `node_modules/@types`.

In this case, `@types/node@16.x` is installed at the root `node_modules` like below:

```
node_modules
- @types
-- node@16.x
- ts34
index.ts
```

Then `typescript` tries to use the type definitions of `@types/node@16.x`. It causes a failure because of the assertion function syntax. (Test it with `npx tsc ./index.ts`)

----

When we add `@types/node@14.14.31` as a dependency to `tst` like below:

```
"devDependencies": {
  "ts34": "sainthkh/ts34",
  "@types/node": "14.14.31"
}
```

It works because `@types/node@14.x` is installed at the root of `/node_modules/@types` and `@types/node@16.x` is installed at the `/node_modules/ts34/node_modules/@types` like below:

```
node_modules
- @types
-- node@14.x
- ts34
-- node_modules
--- @types
---- node@16.x
index.ts
```

Then, when we compile `index.ts`, TypeScript doesn't emit an error because it doesn't use types under `ts34/node_modules/@types`. 

It's how TypeScript handles types and version collisions. 

----

The best rule might be one of these:

* ignore the break and tell them to update TypeScript or install their own `@types/node`. 
* or update the `@types/node` and `TypeScript` requirement at every major update of Cypress. (This doesn't happen that much because Cypress major update happens 3-4 a year, but `@types/node` update happens only once a year.)